source/ice_api.c

```diff
@@ -0,0 +1,1676 @@
+            {
+                *ppCandidatePair = &( pIceAgent->iceCandidatePairs[ i ] );
+                found = true;
+            }
+        }
+    }
+
+    return found;
+}
+/*------------------------------------------------------------------------------------------------------------------*/
+
+/* Ice_IsSameIpAddress - This API is called internally to check if two IPAddresses are same. */
+
+static bool Ice_IsSameIpAddress( StunAttributeAddress_t * pAddr1,
+                                 StunAttributeAddress_t * pAddr2,
+                                 bool checkPort )
+{
+    bool ret;
+    uint32_t addrLen;
+
+    if( ( pAddr1 == NULL ) || ( pAddr2 == NULL ) )
+    {
+        ret = false;
+    }
+    else
+    {
+        addrLen = ICE_IS_IPV4_ADDR( *pAddr1 ) ? STUN_IPV4_ADDRESS_SIZE : STUN_IPV6_ADDRESS_SIZE;
+
+        ret = ( pAddr1->family == pAddr2->family && memcmp( pAddr1->address,
+                                                            pAddr2->address,
+                                                            addrLen ) == 0 && ( !checkPort || pAddr1->port == pAddr2->port ) );
+    }
+
+    return ret;
+}
+/*------------------------------------------------------------------------------------------------------------------*/
+
+/* Ice_CreateIceAgent - The application calls this API for starting a new ICE agent. */
+
+IceResult_t Ice_CreateIceAgent( IceAgent_t * pIceAgent,
+                                char * pLocalUsername,
+                                char * pLocalPassword,
+                                char * pRemoteUsername,
+                                char * pRemotePassword,
+                                char * pCombinedUsername,
+                                TransactionIdStore_t * pBuffer,
+                                Ice_ComputeRandom computeRandomFunction,
+                                Ice_ComputeCrc32 computeCRC32Function,
+                                Ice_ComputeHMAC computeHMACFunction )
+{
+    IceResult_t retStatus = ICE_RESULT_OK;
+    int i;
+
+    if( ( pIceAgent == NULL ) ||
+        ( pLocalPassword == NULL ) ||
+        ( strlen( pLocalPassword ) > ( ICE_MAX_CONFIG_CREDENTIAL_LEN + 1 ) ) ||
+        ( pLocalUsername == NULL ) ||
+        ( strlen( pLocalUsername ) > ( ICE_MAX_CONFIG_USER_NAME_LEN + 1 ) ) ||
+        ( pRemotePassword == NULL ) ||
+        ( strlen( pRemotePassword ) > ( ICE_MAX_CONFIG_CREDENTIAL_LEN + 1 ) ) ||
+        ( pRemoteUsername == NULL ) ||
+        ( strlen( pRemoteUsername ) > ( ICE_MAX_CONFIG_USER_NAME_LEN + 1 ) ) ||
+        ( pCombinedUsername == NULL ) ||
+        ( strlen( pCombinedUsername ) > ( ( ICE_MAX_CONFIG_USER_NAME_LEN + 1 ) << 1 ) ) ||
+        ( pBuffer == NULL ) )
+    {
+        retStatus = ICE_RESULT_BAD_PARAM;
+    }
+
+    if( retStatus == ICE_RESULT_OK )
+    {
+        memset( pIceAgent,
+                0,
+                sizeof( IceAgent_t ) );
+
+        strcpy( pIceAgent->localUsername,
```

Comment:
can we use memcpy for consistency with size provided.

Reply:
updated.